src/mod_stream_management.erl

```diff
@@ -128,17 +293,74 @@ register_smid(SMID, SID) ->
               {error, Reason}
     end.
 
-get_sid(SMID) ->
-    case mnesia:dirty_read(sm_session, SMID) of
-        [#sm_session{sid = SID}] ->
-            [SID];
-        [] ->
-            []
+register_stale_smid_h(SMID, H) ->
+    case gen_mod:get_module_opt(?MYNAME, ?MODULE, stale_h, {false, []}) of
+        {false, []} -> ok;
+        {true, _} ->
+            try
+                Stamp = now_to_seconds(erlang:timestamp()),
+                mnesia:sync_dirty(fun mnesia:write/1,
+                                  [#stream_mgmt_stale_h{
+                                      smid = SMID, h = H, stamp = Stamp}])
+            catch exit:Reason ->
+                      {error, Reason}
+            end
+    end.
+
+remove_stale_smid_h(SMID) ->
+    case gen_mod:get_module_opt(?MYNAME, ?MODULE, stale_h, {false, []}) of
+        {false, []} -> ok;
+        {true, _} -> mnesia:dirty_delete(stream_mgmt_stale_h, SMID)
     end.
 
+%%
+%% gen_server
+start_cleaner(Host, Opts) ->
+    ChildSpec = {?MODULE,
+                 {?MODULE, start_link, [Host, Opts]},
+                 permanent, 5000, worker, [?MODULE]},
+    ejabberd_sup:start_child(ChildSpec).
+
+stop_cleaner(Proc) ->
+    ejabberd_sup:stop_child(Proc).
+
+start_link(Host, Opts) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Host, Opts], []).
+
+init([_Host, GCOpts]) ->
+    RepeatAfter = proplists:get_value(gc_repeat_after, GCOpts, 1800),
+    GeriatricAge = proplists:get_value(gc_geriatric, GCOpts, 3600),
+    State = #state{gc_repeat_after = RepeatAfter,
+                   gc_geriatric = GeriatricAge},
+    {ok, State, RepeatAfter}.
+
+handle_call(_Request, _From, State) ->
+    {reply, ok, State}.
+
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(timeout, #state{gc_repeat_after = RepeatAfter,
+                            gc_geriatric = GeriatricAge} = State) ->
+    TimeToDie = now_to_seconds(erlang:timestamp()) + GeriatricAge,
+    ets:select_delete(
+      stream_mgmt_stale_h,
+      ets:fun2ms(
+        fun(#stream_mgmt_stale_h{stamp=S}) when S < TimeToDie -> true end
+       )
+     ),
+    {noreply, State, RepeatAfter};
+handle_info(_Info, #state{gc_repeat_after = RepeatAfter,
+                          gc_geriatric = _GeriatricAge} = State) ->
+    {noreply, State, RepeatAfter}.
+
 %%
 %% Helpers
 %%
+-spec now_to_seconds(erlang:timestamp()) -> non_neg_integer().
+now_to_seconds({MegaSecs, Secs, _MicroSecs}) ->
```

Comment:
How about renaming `now_to_seconds` to `timestamp_to_seconds`?
Btw it looks very similar to http://erlang.org/doc/man/calendar.html#time_to_seconds-1 ;)

Reply:
Rename, cool. Similarity, that's saying about time from midnight, which doesn't work, it would break if we generate a key at 23:59 and garbage collect at 00:01, but having a geriatric age of two hours 😛